visualization/app/codeCharta/ui/legendPanel/legendMarkedPackages/legendMarkedPackages.component.tsx

```diff
@@ -0,0 +1,63 @@
+import "./legendMarkedPackages.component.scss"
+import { IRootScopeService } from "angular"
+import { CodeMapActionsService } from "../../codeMap/codeMap.actions.service"
+import { MarkedPackage } from "../../../codeCharta.model"
+import { MarkedPackagesSubscriber, MarkedPackagesService } from "../../../state/store/fileSettings/markedPackages/markedPackages.service"
+import { hasValidHexLength, normalizeHex } from "../../customColorPicker/colorHelper"
+
+type Path = string
+interface MarkedPackageMap {
+	[color: string]: Path[]
+}
+
+export class LegendPanelController implements MarkedPackagesSubscriber {
+	private _viewModel: {
+		packageLists: MarkedPackageMap
+	} = {
+		packageLists: {}
+	}
+
+	private _colorPickerScopes = {}
+
+	constructor(private $rootScope: IRootScopeService, private $scope, private codeMapActionsService: CodeMapActionsService) {
+		this.$scope._colorPickerScopes = {}
+		MarkedPackagesService.subscribe(this.$rootScope, this)
+	}
+
+	onMarkedPackagesChanged(markedPackages: MarkedPackage[]) {
+		console.log(this.$scope, this._colorPickerScopes)
+		this._viewModel.packageLists = markedPackages.reduce((acc, markedPackage) => {
+			if (!acc.hasOwnProperty(markedPackage.color)) acc[markedPackage.color] = []
+			acc[markedPackage.color].push(markedPackage.path)
+			return acc
+		}, {})
+
+		this._colorPickerScopes = Object.entries(this._viewModel.packageLists).reduce((acc, [color, paths]) => {
+			acc[paths[0]] = color
+			return acc
+		}, {})
```

Comment:
Note (does not need a follow-up): this could be combined with the above reduce. It's always the first path and as such it's possible to just add something to the `if` statement as in `colorPickerScopes[markedPackage.path] = markedPackage.color`.

Reply:
Yes. But as performance is absolutely no problem here, I decided to be generous and spend the separate two function names for better readability of the code.